text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+API of `LayoutVerified`, which can read off the front of a slice: under the
+covers, it is splitting the pointer, ensuring that it has exclusive access. This
+pattern may cover *some* of our needs, but not all of them, and even where it
+was feasible it would add significant complexity. With the top-level
+tables, for instance, we could verify when loading the font that the reported
+offsets for the tables were all disjoint, and then we could split the underlying
+data into chunks, each of which would be independently mutable; we could either
+do this explicitly (allocating an array of disjoint slices) or via some other
+mechanism, similar to a `RefCell`, where we would keep track at runtime of which regions
+were currently borrowed.
+
+This would work (although it would add significant complexity) in cases where we
+had foreknowledge of the position and size of an object, but I do not believe
+this could work when dealing with tables that host offsets, or at least not in
+the general case. The problem with offsets is that they are not verified ahead
+of time, and they often point to objects that do not have a known size. This
+means there is really no way to know exactly how much memory a table at a given
+offset is going to access. While it is possible to imagine complicated runtime
+schemes for tracking this, it would definitely require allocating additional
+data structures, and would significantly degrade the ergonomics of the API, and
+significantly increase the learning curve.
+
+I've been thinking about this problem quite a bit, and my opinion it is evidence
+of a fundamental tension in the design goals of this project.
+
+#### deciding on goals
+
+I believe that we currently have two conflicting goals. One of the rationale for
+this project is concern about the difficulty of getting developers up to speed
+on HarfBuzz, and how hard it is to add features there without accidentally
+introducing bugs. There is a hope that Rust's safety guarantees can give us a
+code base that is easier to modify and hack on; however there is *also* a desire
+to end up with code that is more-or-less equivalent to the existing C++.
+
+The basic tension is this: the ergonomics and safety advantages that Rust offers
+work by explicitly preventing certain behaviour, such as the aliasing of mutable
+variables. We cannot translate idiomatic C++ directly into Rust and get safety;
+we get safety by ruling out certain patterns.
+
+I think it is really worth stepping back and considering our goals. We
+would like to have a tool for authoring & manipulating font files, and we would
+like it to be a) fast and b) easy to use. This is something that we can achieve,
+but I do not think the result is going to look like it might look in C++. My
+feeling is that too much of our thinking (or at least my thinking!) so far has
+been focused on porting an existing design to Rust, and that this may be
+misguided.
```

Comment:
I don't think we want to translate idiomatic C++ to Rust code that "looks the same." I think I would say we want to thoughtfully aim to mirror the algorithms and general approach existing best in class implementations (e.g. HarfBuzz for shaping and subsetting, FontTools for font binary assembly) take, and probably to actually test that the Rust version matches the results given by the existing implementation. We may choose to deliberately deviate but I would expect that to be the exception not the rule. That enables us to take full advantage of existing experts and to avoid becoming endlessly bogged down in how to approach things because we have a default answer: the way we do in the existing implementation unless there is a very good reason to do otherwise.

@garretrieger followed more or less this approach with great success in producing hb-subset by testing it against FontTools. Migrating existing software becomes much easier when the replacement produces the same results, just faster (or safer or w/e).

Reply:
That makes me want to build a subset of subsetting - must produce same bytes as hb-subset and we can compare speed, memory use, etc as well - as one of our early explorations. Do you think that might make sense?